<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>Usage guideline - CusProSe documentation</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Usage guideline";
        var mkdocs_page_input_path = "ihb_usage.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="index.html" class="icon icon-home"> CusProSe documentation
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="index.html">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Iterhmmbuild</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="ihb_introduction.html">Overall procedure</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="ihb_usage.html">Usage guideline</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#quick-start-examples">Quick start examples</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#build-a-single-hmm-profile">Build a single HMM profile</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#build-an-hmm-profile-database">Build an HMM profile database</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#command-line-and-parameters">Command line and parameters</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#output-of-iterhmmbuild">Output of IterHMMBuild</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#output-from-the-generation-of-a-single-hmm-profile">Output from the generation of a single HMM profile</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#output-from-the-generation-of-an-hmm-profile-database">Output from the generation of an HMM profile database</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Prosecda</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="psd_introduction.html">Overall procedure</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="psd_usage.html">Usage guideline</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CusProSe documentation</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html" class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>Iterhmmbuild &raquo;</li>
      <li>Usage guideline</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="iterhmmbuild-usage-guideline">IterHMMBuild usage guideline<a class="headerlink" href="#iterhmmbuild-usage-guideline" title="Permanent link">#</a></h1>
<div class="admonition tip">
    <p class="first admonition-title">
        Tip
    </p>
    <p>
    To guide the user, input data examples can be found in <code>cusProSe/iterhmmbuild/datas/</code>:
    <pre style="line-height: 15px;"><b>cusProSe/iterhmmbuild/datas/</b>
    ├── inputs
    │   ├── A.fa
    │   ├── KS.fa
    │   └── PP.fa
    └── mgg_70-15_8.fasta
    </pre>

    All fasta files in the <code>inputs/</code> directory contain sequences of three different protein domains. There is also the <i>magnaporthe orizae</i> proteome (<code>mgg_70-15_8.fasta</code>) that will be used as the protein database for the examples below.
    </p>
</div>

<h2 id="quick-start-examples">Quick start examples<a class="headerlink" href="#quick-start-examples" title="Permanent link">#</a></h2>
<h4 id="build-a-single-hmm-profile">Build a single HMM profile<a class="headerlink" href="#build-a-single-hmm-profile" title="Permanent link">#</a></h4>
<div>
Command to build a single HMM profile from the <i>magnaporthe orizae</i> proteome and representative of the A domain sequences:

<pre><code class="language-bash">iterhmmbuild -fa inputs/A.fa -protdb mgg_70-15_8.fasta
</code></pre>

</div>

<h4 id="build-an-hmm-profile-database">Build an HMM profile database<a class="headerlink" href="#build-an-hmm-profile-database" title="Permanent link">#</a></h4>
<div>
Command to build an HMM profile database specific to the <i>magnaporthe orizae</i> proteome with each profiles representative of the domain sequences present in the directory `inputs/`:

<pre><code class="language-bash">iterhmmbuild -fa inputs/ -protdb mgg_70-15_8.fasta
</code></pre>

</div>

<div class="admonition note">
    <p class="first admonition-title">Note</p>
    <p>
    Please note that the user can also create an HMM profile database from a set of individual HMM profiles through the command <code class="io">create_hmmdb</code>. Those HMM profiles must be placed in a unique directory that is given as an input.
    </p>
    <p>
    For instance, let's say we have the following directory containing three HMM profiles:
        <pre style="line-height: 15px;"><b>my_hmm_dir/</b>
        ├── A.hmm
        ├── KS.hmm
        └── PP.hmm
        </pre>
    Then the following command will generate an HMM profile database (that is a simple concatenation of the three HMM profiles) called <code class="io">mydb.hmm</code> in an output directory named <code class="io">databases/</code>: 
    </p>
    <p>
    <code class="inside">create_hmmdb -hmmdir my_hmm_dir/ -dbname mydb.hmm -outdir databases/</code>
    </p>
</div>

<h4 id="command-line-and-parameters">Command line and parameters<a class="headerlink" href="#command-line-and-parameters" title="Permanent link">#</a></h4>
<p>As illustrated by the commands above, two arguments are mandatory for IterHMMBuild: <code>-fa</code> and <code>-protdb</code>.</p>
<ul class="myul">
<li>
The input following <code>-fa</code> can be either a fasta file with at least one protein sequence OR a directory location where multiple individual fasta files are stored. The former can be used when the user wants to build a single HMM profile representative of the sequence(s) given as input. The latter should be used when the user wants to build a set of HMM profiles concatenated into an HMM profile database, with each HMM profile being representative of each related fasta files present in the directory given as input.
</li>
<li>
The input following <code>-protdb</code> is a fasta file of the protein database used to enrich initial protein sequence(s) of interest.
</li>
</ul>

<p>Help about the usage of IterHMMBuild and its parameters can be shown with the following command: <code>iterhmmbuild -h</code></p>
<pre class="parameters">usage: iterhmmbuild [-h] -fa [FA] -protdb [PROTDB] [-name [NAME]] [-out [OUT]] [-id ID] 
                    [-cov COV] [-cval CVAL] [-ival IVAL] [-acc ACC] [-delta DELTA]
                    [-maxcount MAXCOUNT]

Iterative building of hmm profiles

optional arguments:
  -h, --help          show this help message and exit
  -fa [FA]            Fasta file of sequence(s) used as first seed or directory containing such files
  -protdb [PROTDB]    Sequences used to learn the hmm profile (fasta format)
  -name [NAME]        Name for the HMM profile (fasta name by default).
  -out [OUT]          Output directory
  -id ID              Sequence identity threshold to remove redundancy in seeds&apos;sequences (0.9)
  -cov COV            Minimum percentage of coverage alignment between hmm hit and hmm profile (0.0)
  -cval CVAL          HMMER conditional e-value cutoff (0.01)
  -ival IVAL          HMMER independant e-value cutoff (0.01)
  -acc ACC            HMMER mean probability of the alignment accuracy between each residues of the target and the 
                      corresponding hmm state (0.6)
  -delta DELTA        Convergence criteria: difference in the number of sequences found between two consecutive iterations            
                      to consider a non-significant change between between two consecutive iterations (1)
  -maxcount MAXCOUNT  Convergence criteria: maximum number of times a non-significant change (conv_delta) is accepted before
                      considering a convergence (3)
</pre>

<h2 id="output-of-iterhmmbuild">Output of IterHMMBuild<a class="headerlink" href="#output-of-iterhmmbuild" title="Permanent link">#</a></h2>
<p>After running IterHMMBuild an output directory will be generated in the following generic format: 
<code>iterhmmbuild_year-month-day_hour-min-sec/</code></p>
<h4 id="output-from-the-generation-of-a-single-hmm-profile">Output from the generation of a single HMM profile<a class="headerlink" href="#output-from-the-generation-of-a-single-hmm-profile" title="Permanent link">#</a></h4>
<p>The output directory generated from the <a href="#build-a-single-hmm-profile">command run in the quick start examples</a> will have the following architecture:</p>
<pre><b>iterhmmbuild_2020-10-29_13-13-04/</b>
├── A.hmm
├── A_seed.clw
├── A_seed.fa
├── info.log
├── <b>iter_1/</b>
├── <b>iter_2/</b>
├── <b>...</b>
└── <b>iter_6/</b>
</pre>

<p>The three main files of interest are:</p>
<table class="t-description">
    <tr>
        <td class="t-data"><b>A.hmm</b></td> 
        <td class="t-data">Final HMM profile</td> 
    </tr>
    <tr>
        <td class="t-data"><b>A.seed.clw</b></td> 
        <td class="t-data">Final sequences used to build A.hmm</td> 
    </tr>
    <tr>
        <td class="t-data"><b>A.seed.fa</b></td> 
        <td class="t-data">Multiple alignment (clustal W format) of A_seed.fa</td> 
    </tr>
</table>

<p><code>info.log</code> is a log summary of the computation. The subdirectories <code>iter_i/</code> contain files obtained at each iteration and are described in section <a href="./ihb_introduction.html#outputs">Overall procedure</a>.</p>
<h4 id="output-from-the-generation-of-an-hmm-profile-database">Output from the generation of an HMM profile database<a class="headerlink" href="#output-from-the-generation-of-an-hmm-profile-database" title="Permanent link">#</a></h4>
<p>The output directory generated from the <a href="#build-an-hmm-profile-database">command run in the quick start examples</a> will be a list of subdirectories such as the output described above. You will find at its root the file <code>hmm_database.hmm</code>, a concatenation of the HMM profiles of protein domains used as inputs. </p>
<pre><b>iterhmmbuild_2021-03-02_12-39-38</b>
├── hmm_database.hmm
├── info.log
├── <b>A</b>
│   ├── A.hmm
│   ├── A_seed.clw
│   └── A_seed.fa
├── <b>KS</b>
│   ├── KS.hmm
│   ├── KS_seed.clw
│   └── KS_seed.fa
└── <b>PP</b>
    ├── PP.hmm
    ├── PP_seed.clw
    └── PP_seed.fa
</pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="ihb_introduction.html" class="btn btn-neutral float-left" title="Overall procedure"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="psd_introduction.html" class="btn btn-neutral float-right" title="Overall procedure">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="ihb_introduction.html" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="psd_introduction.html" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme_extra.js" defer></script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
